### PR TITLE
wip: implement service for initial indexing of existing chats

### PR DESCRIPTION
**Pull request by AlexUstinov**

### CODE REVIEW

**Comment by AlexUstinov** [comment]
I didn't test it yet, so there might be some errors there.

File: src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializer.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+internal class ChatIndexInitializer(
+    IServiceProvider services,
+    ShardScheme shardScheme,
+    IShardIndexResolver<string> shardIndexResolver,
+    IChatIndexInitializerShard chatIndexInitializerShard,
+    ILogger<ChatIndexInitializer> log
+) : ShardWorker(services, shardScheme, nameof(ChatIndexInitializer)), IChatIndexInitializer
+{
+    private record DummyEvent() : IHasShardKey<string>
```
Comment (andrew-kurochkin) [inline src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializer.cs:18]: - Not a blocker - 
Note: this must be solved with commands sent to a correct sharding scheme

File: src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializerShard.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    private async Task UpdateCursor(CancellationToken cancellationToken)
+    {
+        await Task.Yield();
+
+        var lastEventCount = Volatile.Read(ref _eventCount);
+        while (!cancellationToken.IsCancellationRequested) {
+            try {
+                await clock.Delay(TimeSpan.FromSeconds(30), cancellationToken).ConfigureAwait(false);
```
Comment (andrew-kurochkin) [inline src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializerShard.cs]: - Not a blocker -
Would be nice to move this into config with a default of 30sec.

File: src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializerShard.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            while(true) {
+                try {
+                    var job = new MLSearch_TriggerChatIndexing(chatId);
+                    await commander.Call(job, cancellationToken).ConfigureAwait(false);
+                    _scheduledJobs[chatId] = (version, clock.Now.EpochOffset.Ticks);
+                    break;
+                }
+                catch(Exception e) when (e is not OperationCanceledException) {
+                    log.LogError(e, "Failed to schedule an indexing job for chat #{ChatId}.", chatId);
+                    await clock.Delay(TimeSpan.FromSeconds(10), cancellationToken).ConfigureAwait(false);
```
Comment (andrew-kurochkin) [inline src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializerShard.cs]: - Not a blocker - 
Would be nice to have it in a config (retry timeout)

File: src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializerShard.cs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            try {
+                batch = await chats
+                    .ListChanged(minVersion, long.MaxValue, ChatId.None, BatchSize, cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            catch(Exception e) when (e is not OperationCanceledException) {
+                log.LogError(e,
+                    "Failed to load a batch of chats of length {Len} in the version range from {MinVersion} to infinity.",
+                    BatchSize, minVersion);
+                await clock.Delay(TimeSpan.FromSeconds(10), cancellationToken).ConfigureAwait(false);
```
Comment (andrew-kurochkin) [inline src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializerShard.cs]: Same here

File: src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializerShard.cs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                continue;
+            }
+            if (batch.Count==0) {
+                await Task.Delay(TimeSpan.FromMinutes(1), cancellationToken).ConfigureAwait(false);
```
Comment (andrew-kurochkin) [inline src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializerShard.cs]: I believe it should be configurable

**Review by andrew-kurochkin — APPROVED** [review]

File: src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializer.cs (hunk 1 above)
Comment (AlexUstinov) [reply src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializer.cs:18]: I agree with this, so I added todo item for the corresponding refactoring.

**Pull request merged**